### PR TITLE
Add logic for finding `MentionsAuthorReference`s

### PR DESCRIPTION
**Pull request by decentralion**
A `MentionsAuthorReference` is created when a post mentions a user, and
that user has authored at least one post in the same thread. Then there
is a `MentionsAuthorReference` from the post to the other posts by that
author.

For context, see the docstrings in `mentionsAuthorReference.js`, and
see #804.

Test plan:
Thorough unit tests have been added, which test the entire pipeline,
from ingesting the data via GitHub's graphql responses, through to
detecting the references. Edge cases such as self-reference and
multi-reference are tested.

Thanks to @wchargin for help writing this commit.

### CODE REVIEW

File: src/plugins/github/heuristics/mentionedAuthorReference.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import * as RV from "../relationalView";
+import * as N from "../nodes";
+
+// MentionedAuthorReferences are inferred edges between posts.
+// If a post contains an @-mention of a user, and that user has authored any posts
+// in the same thread, then we create MentionedAuthorReferences from the mentioning
+// post to any posts authored by the mentioned author.
+// A concrete example: Suppose I write a post, and I @-mention you. Then my post will
+// have MentionedAuthorReferences pointing to any posts that you write. The inference
+// is that because I mentioned you by name, my post probably values your posts (either because
```
Comment (wchargin) [inline src/plugins/github/heuristics/mentionedAuthorReference.js]: nit: this line has length 94; please wrap (preferably at 72)

File: src/plugins/github/heuristics/mentionedAuthorReference.test.js [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    /** The test cases below are organized so that for every
+     * post in the thread rooted at Issue #2, we expect all
+     * of the new MentionedAuthorReferences that were created
+     * as a result of that post (whether because that post is
+     * the src or the dst). At the end, we verify that every
+     * MentionedAuthorReference is accounted for. */
+
+    // If an author references themself, there will be a self-directed
+    // MentionedAuthorReference.
+    expectReferences("issue:2", [
```
Comment (wchargin) [inline src/plugins/github/heuristics/mentionedAuthorReference.test.js]: As these strings translate to test case names, can we make them describe
the expected behavior at the domain level (e.g., “a self-reference on a
comment”) rather than the data-specific level (“comment:1_1”)?

File: src/plugins/github/heuristics/mentionedAuthorReference.test.js [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    ]);
+    expectReferences("comment:2_2", [
+      {
+        src: commentAddress(2, 2),
+        dst: commentAddress(2, 1),
+        who: users.holoPearl,
+      },
+      // Because Steven self-referenced, that post references all of Steven's
+      // other posts.  (If people use this for gaming, we can reconsider, but
+      // it shoudl be fairly obvious)
```
Comment (wchargin) [inline src/plugins/github/heuristics/mentionedAuthorReference.test.js]: shoudl

File: src/plugins/github/heuristics/mentionedAuthorReference.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  // there might be multiple pointers from the same src to the same dst
+  +who: N.UserlikeAddress,
+|};
+
+export function* findMentionedAuthorReferences(
+  view: RV.RelationalView
+): Iterator<MentionedAuthorReference> {
+  function* issuesAndPulls() {
+    yield* view.issues();
+    yield* view.pulls();
```
Comment (wchargin) [inline src/plugins/github/heuristics/mentionedAuthorReference.js]: If you delete `yield* view.pulls()`, the tests all pass.

Can we make the tests just a wee bit stronger?

File: src/plugins/github/heuristics/mentionedAuthorReference.test.js (hunk 2 above)
Comment (decentralion) [reply src/plugins/github/heuristics/mentionedAuthorReference.test.js]: We could, but at the expense of readability and maintainability. When writing these tests, it's important to have a clear place where every individual reference should live. As written, there's a simple system: we place each reference with the {src} or {dst} of that reference, choosing whichever one was chronologically last. If I re-write them to have descriptive names, it won't be unambiguous which block should own any particular reference, making the code harder to maintain.

File: src/plugins/github/heuristics/mentionedAuthorReference.js (hunk 4 above)
Comment (decentralion) [reply src/plugins/github/heuristics/mentionedAuthorReference.js]: Ok.

Just a wee bit, though.

File: src/plugins/github/heuristics/mentionsAuthorReference.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
+// @flow
+
+import * as MapUtil from "../../../util/map";
+import * as RV from "../relationalView";
+import * as N from "../nodes";
+
+// MentionsAuthorReferences are inferred edges between posts.  If a post
+// contains an @-mention of a user, and that user has authored any posts
+// in the same thread, then we create MentionsAuthorReferences from the
+// mentioning post to any posts authored by the mentioned author.  A
```
Comment (wchargin) [inline src/plugins/github/heuristics/mentionsAuthorReference.js]: Mind being a bit more explicit and writing “any posts authored by the
mentioned author within the same thread”? The fact that this is
thread-scoped is only hinted at in the introductory sentence.

File: src/plugins/github/heuristics/mentionsAuthorReference.js [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+  +src: N.TextContentAddress,
+  // The post authored by the person mentioned in the src
+  +dst: N.TextContentAddress,
+  // The person that was mentioned in src.  Storing `who` is necessary
+  // because in the case of paired authorship, there might be multiple
+  // pointers from the same src to the same dst
+  +who: N.UserlikeAddress,
+|};
+
+export function* findMentionsAuthorReferences(
```
Comment (wchargin) [inline src/plugins/github/heuristics/mentionsAuthorReference.js]: I like that this and its return type are the only two exports. Good encapsulation.
